pkg/csi/service/vanilla/controller.go

```diff
@@ -1555,26 +1556,38 @@ func (c *controller) ListVolumes(ctx context.Context, req *csi.ListVolumesReques
 					string(cnstypes.QuerySelectionNameTypeVolumeType),
 				},
 			}
-			// Select only the volume type.
-			cnsQueryResult, err = c.manager.VolumeManager.QueryAllVolume(ctx, queryFilter, querySelection)
-			if err != nil {
-				return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
-					"queryVolume failed on Cluster ID %q with err = %+v ", cfg.Global.ClusterID, err)
+			// For multi-VC configuration, query volumes from all vCenters
+			if multivCenterCSITopologyEnabled {
+				for _, volumeManager := range c.managers.VolumeManagers {
+					cnsQueryResult, err = volumeManager.QueryAllVolume(ctx, queryFilter, querySelection)
+					if err != nil {
+						return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+							"queryVolume failed on Cluster ID %q with err = %+v ", cfg.Global.ClusterID, err)
+					}
+					cnsVolumes = append(cnsVolumes, cnsQueryResult.Volumes...)
+				}
+			} else {
+				cnsQueryResult, err = c.manager.VolumeManager.QueryAllVolume(ctx, queryFilter, querySelection)
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"queryVolume failed on Cluster ID %q with err = %+v ", cfg.Global.ClusterID, err)
+				}
+				cnsVolumes = cnsQueryResult.Volumes
 			}
 		}
 
 		// Step 3: If the difference between number of K8s volumes and CNS volumes is greater than threshold,
 		// fail the operation, as it can result in too many attach calls.
-		if len(volIDsInK8s)-len(cnsQueryResult.Volumes) > cfg.Global.ListVolumeThreshold {
+		if len(volIDsInK8s)-len(cnsVolumes) > cfg.Global.ListVolumeThreshold {
 			log.Errorf("difference between number of K8s volumes: %d, and CNS volumes: %d, is greater than "+
-				"threshold: %d, and completely out of sync.", len(volIDsInK8s), len(cnsQueryResult.Volumes),
+				"threshold: %d, and completely out of sync.", len(volIDsInK8s), len(cnsVolumes),
```

Comment:
You could set the `cnsQueryResult` to nil, which can reduce the memory usage of the driver. Please evaluate if this causes any other issues.